File: CS-288/homework-07/nasdaq.py
import sys
import mysql.connector
from xml.dom.minidom import parse

username = sys.argv[2]
password = sys.argv[3]
xhtmlFile = sys.argv[1]

timestamp = xhtmlFile.split("/")
timestamp = timestamp[len(timestamp) - 1].split(".")[0]


# inserts a row of values into the table
def insert(cursor, values):
    query = "INSERT INTO nasdaq(timestamp,exchange,symbol,company,volume,price,percentChange) VALUES (%s,%s,%s,%s,%s,%s,%s)"
    cursor.execute(query, values)


# returns the text from the given nodeList
def getText(nodeList):
    for node in nodeList.childNodes:
        if node.nodeType == node.TEXT_NODE:
            return str(node.nodeValue)


# returns a tuple of values
def getRowData(tr):
    td = tr.getElementsByTagName("td")
    anchor = td[1].getElementsByTagName("a")[0]
    anchorText = getText(anchor).split("(")
    company = anchorText[0].strip()
    symbol = anchorText[1][: len(anchorText[1]) - 2]
    volume = getText(td[2]).replace(",", "")
    price = getText(td[3]).replace(",", "").replace("$", "")
    percentChange = getText(td[4])
    return (timestamp, "Nasdaq", symbol, company, volume, price, percentChange)


if len(sys.argv) < 3:
    print("missing args")
    exit(1)


try:
    cnx = mysql.connector.connect(
        host="localhost", user=username, password=password, database="demo"
    )
    cursor = cnx.cursor()

    document = parse(xhtmlFile)
    table = document.getElementsByTagName("table")[2]
    tableRows = table.getElementsByTagName("tr")[1:]

    for tr in tableRows:
        values = getRowData(tr)
        insert(cursor, values)

    # commit and close
    cnx.commit()
    cursor.close()
except mysql.connector.Error as err:
    print(err)
finally:
    try:
        cnx
    except NameError:
        pass
    else:
        cnx.close()

File: CS-288/homework-07/nasdaq.php
<html>

<head>
    <title>Nasdaq Data</title>
</head>

<body>
    <table align="center" border="1">

        <?php
        
        // sorting is handled based on query string params
        $queryString = $_GET ? $_SERVER['QUERY_STRING'] : 'orderBy=symbol&sort=ASC';

        // log the query string to the console
        echo '<script> console.log("'.$queryString.'") </script>';

        echo '<tr style="text-align: center;">';
        echo '<td>exchange</td>';
        
        if ($queryString == 'orderBy=symbol&sort=ASC') {
            echo '<td><a href="?orderBy=symbol&sort=DESC">symbol</a></td>';
        } else {
            echo '<td><a href="?orderBy=symbol&sort=ASC">symbol</a></td>';
        }
        
        if ($queryString == 'orderBy=company&sort=ASC') {
            echo '<td><a href="?orderBy=company&sort=DESC">company</a></td>';
        } else {
            echo '<td><a href="?orderBy=company&sort=ASC">company</a></td>';
        }

        if ($queryString == 'orderBy=volume&sort=ASC') {
            echo '<td><a href="?orderBy=volume&sort=DESC">volume</a></td>';
        } else {
            echo '<td><a href="?orderBy=volume&sort=ASC">volume</a></td>';
        }

        if ($queryString == 'orderBy=price&sort=ASC') {
            echo '<td><a href="?orderBy=price&sort=DESC">price</a></td>';
        } else {
            echo '<td><a href="?orderBy=price&sort=ASC">price</a></td>';
        }

        if ($queryString == 'orderBy=percentChange&sort=ASC') {
            echo '<td><a href="?orderBy=percentChange&sort=DESC">change</a></td>';
        } else {
            echo '<td><a href="?orderBy=percentChange&sort=ASC">change</a></td>';
        }

        echo '</tr>';

        $cnx = new mysqli('localhost', 'user', 'password', 'db');

        if ($cnx->connect_error) {
            die('Connection failed: ' . $cnx->connect_error);
        }

        // default sort by symbol
        $validCols = array('symbol', 'company', 'volume', 'price', 'percentChange');
        $orderBy = isset($_GET['orderBy']) && in_array($_GET['orderBy'], $validCols) ? $_GET['orderBy'] : 'symbol';

        // default sort ASC
        $sortyBy = (isset($_GET['sort']) && $_GET['sort'] == 'ASC' || $_GET['sort'] == 'DESC') ? $_GET['sort'] : 'ASC';

        $query = "SELECT * FROM table ORDER BY $orderBy $sortyBy";
        $cursor = $cnx->query($query);

        while ($row = $cursor->fetch_assoc()) {
            echo '<tr>';
            echo '<td>' . $row['exchange'] . '</td><td>' . $row['symbol'] . '</td><td>' . $row['company'] . '</td><td align="right">' . $row['volume'] . '</td><td align="right">' . $row['price'] .'</td><td align="right">' . $row['percentChange'] .'</td>';
            echo '</tr>';
        }

        $cnx->close();
        ?>

    </table>
</body>

</html>
